src/utils.rs

```diff
@@ -0,0 +1,280 @@
+use crate::error::SlateDBError;
+use crate::error::SlateDBError::{BackgroundTaskFailed, BackgroundTaskShutdown};
+use std::future::Future;
+
+pub(crate) struct WriteOnceRegister<T: Clone> {
+    rx: tokio::sync::watch::Receiver<Option<T>>,
+    tx: tokio::sync::watch::Sender<Option<T>>,
+}
+
+pub(crate) struct WriteOnceRegisterReader<T: Clone> {
+    rx: tokio::sync::watch::Receiver<Option<T>>,
+}
+
+impl<T: Clone> WriteOnceRegister<T> {
+    pub(crate) fn new() -> Self {
+        let (tx, rx) = tokio::sync::watch::channel(None);
+        Self { rx, tx }
+    }
+
+    pub(crate) fn write(&self, val: T) {
+        self.tx.send_if_modified(|v| {
+            if v.is_some() {
+                return false;
+            }
+            v.replace(val);
+            true
+        });
+    }
+
+    pub(crate) fn reader(&self) -> WriteOnceRegisterReader<T> {
+        WriteOnceRegisterReader {
+            rx: self.rx.clone(),
+        }
+    }
+}
+
+impl<T: Clone> WriteOnceRegisterReader<T> {
+    pub(crate) fn read(&self) -> Option<T> {
+        self.rx.borrow().clone()
+    }
```

Comment:
calling this could cause `await_value` to hang forever if
1. `if let Some(maybe_value) = self.rx.borrow().as_ref()` returns `None`
2. a value appears on the watch
3. `read()` returns something
4. `self.rx.changed()` will wait forever

Simple fix is perhaps to use `wait_for` instead of `changed()` and then return (so no manual loop)

Reply:
That can't happen because `borrow` doesn't mark the value as seen. So `changed` would still return immediately.

That said, I totally missed that there's a `wait_for` method which makes this way simpler. So I'll switch to using that.